### ISSUE TITLE
Remain Memoize cache value when sampling without @ForAll

### ISSUE THREAD

**Issue by mhyeon-lee** [jlink/jqwik#236]
### 1. jqwik engine + @ForAll (norma case)

```java
@Example
void test1(@ForAll @AlphaChars String stringValue) {
    System.out.println(stringValue);
}

@Example
void test2(@ForAll @AlphaChars String stringValue) {
    System.out.println(stringValue);
}
```

![스크린샷 2021-10-04 오후 11 23 23](https://user-images.githubusercontent.com/5206572/135873605-a2358b9e-bec1-47ae-95b0-a60f6f647941.png)

At the end of each `@Property/@Example`, jqwik engine execute the FinishScope for PropertyMethodDescriptor scope.
Memoize has only PropertyMethodDescriptor value and will be removed from the finishScope.
It's a normal case.

### 2. jqwik engine + Arbitrary sample
``` java
@Example
void test1() {
    String stringValue = Arbitraries.strings().alpha().sample();
    System.out.println(stringValue);
}

@Example
void test2() {
    String stringValue = Arbitraries.strings().alpha().sample();
    System.out.println(stringValue);
}
```

![스크린샷 2021-10-04 오후 11 31 19](https://user-images.githubusercontent.com/5206572/135873641-8be981e8-889a-4336-93ab-4a9e31aad4fe.png)

PropertyMethodDescriptor and SAMPLE_STREAM_DESCRIPTOR are stored.
PropertyMethodDescriptor is removed by finishScope, but SAMPLE_STREAM_DESCRIPTOR is not removed and is accumulated in the next '@Property/@Example' test.
The Arbitrary declared within the second test is an object different from the Arbitrary of Memoize's SAMPLE_STREAM_DESCRIPTOR scope and is therefore not reused and accumulated. (leak)

If share the Archive as below, the Memoize Cache of SAMPLE_STREAM_DESCRIPTOR scope will be reused. But still, the new Library continues to accumulate.

```java
private final static Arbitrary<String> stringValueArbitrary = Arbitraries.strings().alpha();

@Example
void test1() {
    String stringValue = stringValueArbitrary.sample();
    System.out.println(stringValue);
}

@Example
void test2() {
    String stringValue = stringValueArbitrary.sample();
    System.out.println(stringValue);
}
```

### 3. JUnit + Arbitrary sample (not jqwik scope)
```java
@Test
void test1() {
    String stringValue = Arbitraries.strings().alpha().sample();
    System.out.println(stringValue);
}

@Test
void test2() {
    String stringValue = Arbitraries.strings().alpha().sample();
    System.out.println(stringValue);
}
```

![스크린샷 2021-10-05 오전 12 09 35](https://user-images.githubusercontent.com/5206572/135876539-78b53eb8-8ab4-4d5c-ae4b-5ae49133c0b3.png)

If not jqwik scope, only SAMPLE_STREAM_DESCRIPTOR scope cache remains in Memoize. These are not reused but accumulated because they are different Arbitrary definitions.

## Discussion

The SAMPLE_STREAM_DESCRIPTOR scope cache is not reused and accumulated in two cases (`2. jqwik engine+Arbitrary sample', `3. J Unit+Arbitrary sample (not jqwik scope)`).
Is there a way to remove this cache value?

(I know that the two cases mentioned are not recommended way. However, I create issues to discuss whether there is a solution.)

**Comment by jlink** [comment]
Case 2 is definitely a bug.
Case 3 works as intended but it's not really what one would like to have.

I'll try to think of a useful fix.

**Comment by jlink** [comment]
Fixed memory leak of case 2. 

Fix available in 1.6.0-SNAPSHOT

**Comment by jlink** [comment]
As for case 3: Stores will now be cleaned up if (and only if) the stream returned by `Arbitrary.sampleStream()` is closed explicity, e.g. through using `try(..) { }`. Here's an example:

```
try (Stream<String> stringStream = Arbitraries.strings().ofLength(5).sampleStream()) {
	stringStream.limit(10).forEach(v -> assertThat(v).hasSize(5));
};
```

Closing the stream for every usage of it is IMO not possible in a consistent way. See https://stackoverflow.com/questions/34753078/register-a-stream-completion-hook for a discussion of the options.

The drawbacks of this solution are:
1. `Arbitrary.sample()` does not clean up after itself. This could be done by implicitly closing the underlying stream but then each call to `sample()` would have the overhead of recreating all generators.
2. Closing across several streams or stream usages is possible but looks ugly since only the last stream may be closed and the others not.

I see two options:
1. Stick with the current suboptimal solution 
2. Offer an API to control the necessary jqwik lifecycle hooks outside of the jqwik engine, e.g.

    ```
    PropertySession session = Jqwik.createSession()
    Arbitrary<..> a1 = ...
    Arbitrary<..> a2 = ...
    
    a1.sample()...
    a2.sample()...
    a1.sample()...
    session.close()
    ```
    This could be part of revealing more jqwik features to use outside the engine.

@mhyeon-lee What do you think?

**Comment by mhyeon-lee** [comment]
@jlink Thank you for your support.
Samplestream processing will help me. `try(..) { }`

> Offer an API to control the necessary jqwik lifecycle hooks outside of the jqwik engine, e.g.

If this life cycle hook is supported, it can be more helpful.

I will implement createSession and close in the JUnit TestExecutionListener.

**Comment by jlink** [comment]
> I will implement createSession and close in the JUnit TestExecutionListener.

You probably wouldn't want to do that for each and every JUnit test, but just for those using jqwik arbitraries.

**Comment by jlink** [comment]
Following up with issue about jqwik sessions: https://github.com/jlink/jqwik/issues/237

**Comment by jlink** [comment]
Feature has been released in 1.6.0
